### DISCUSSION TITLE
jsonable_encoder not using custom_encoder for pydantic v2 model instance

### DISCUSSION THREAD

**Discussion by sehraramiz** [fastapi/fastapi#10252] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import datetime

from fastapi.encoders import jsonable_encoder
from pydantic import BaseModel


class Item(BaseModel):
    dt: datetime


current_dt = datetime.now()
item = Item(dt=current_dt)
item_dict = Item(dt=current_dt).model_dump()

custom_encoder = {datetime: lambda dt: dt.strftime("%Y-%m-%d %H:%M")}

item_data = jsonable_encoder(item, custom_encoder=custom_encoder)
item_dict_data = jsonable_encoder(item_dict, custom_encoder=custom_encoder)

print(item_data)
print(item_dict_data)

assert item_data == item_dict_data
```

### Description

- run this example script.

expected result is that jsonable_encoder use the custom_encoder provided and change the dt field when encoding. based on [this line](https://github.com/tiangolo/fastapi/blob/bfde8f3ef20dc338bbce8a0ff0f4b441c54d5b64/fastapi/encoders.py#L124)  of jsonable_encoder function i see that we are using the custom encoder only for the pydantic V1 instances and not using it for V2.
in pydantic V2 migration guide they mention why they deprecated custom json encoders in model config:

> Due to performance overhead and implementation complexity, we have now deprecated support for specifying json_encoders in the model config

but is there a similar reason that we are not using the custom_encoder in jsonable_encoder for pydantic V2 objects?

in summary i want a test like this to pass:

```python
  def test_encode_custom_encoders_model_pydantic():
      class MyModel(BaseModel):
          dt_field: datetime

      def custom_encoder(dt: datetime) -> str:
          return dt.strftime("%Y-%m-%d %H:%M")

      instance = MyModel(dt_field=datetime(2019, 1, 1, 8, 1, 1))
      encoded_instance = jsonable_encoder(
          instance, custom_encoder={datetime: custom_encoder}
      )
      assert encoded_instance == {"dt_field": "2019-01-01 08:01"}

```

my suggested change is like this:

```diff
diff --git a/fastapi/encoders.py b/fastapi/encoders.py
index 30493697..6a1de46e 100644
--- a/fastapi/encoders.py
+++ b/fastapi/encoders.py
@@ -123,14 +123,14 @@ def jsonable_encoder(
         exclude = set(exclude)
     if isinstance(obj, BaseModel):
         # TODO: remove when deprecating Pydantic v1
-        encoders: Dict[Any, Any] = {}
+        encoders: Dict[Any, Any] = custom_encoder
         if not PYDANTIC_V2:
             encoders = getattr(obj.__config__, "json_encoders", {})  # type: ignore[attr-defined]
             if custom_encoder:
                 encoders.update(custom_encoder)
         obj_dict = _model_dump(
             obj,
-            mode="json",
+            mode="python",
             include=include,
             exclude=exclude,
             by_alias=by_alias,
```

it still passes all the tests fo Pydantic V1/V2 and fixes my problem.

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.11

**Comment by wu-clan** [comment]
Hi, @sehraramiz 

I encountered the same problem when I migrated to pydantic-v2, jsonable_encoder's custom_encoder is invalidm, can you submit a PR for this?

**Reply by sehraramiz** [reply]
hi @wu-clan 
i think this is the intended behaviour of fastapi for now,
based on [this comment](https://github.com/tiangolo/fastapi/blob/36c26677682c4245183912e00fc057c05cc6cf7a/fastapi/encoders.py#L239) they are going to remove the custom encoder when pydantic v1 is deprecated compeletly.
i think this behaviour is because they want to respect the json encoders of pydantic, so when we pass a pydantic model instance with a custom json encoder to jsonable_encoder, prevent jsonable_encoder from overriding expected encodings of pydantic model.
but also based on [this section of pydantic docs](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.json_encoders) the custom json encoder is deprecated in pydantic v2 and i dont know any other way of defining custom json encoders in pydantic v2 (maybe serializers are enough).
so right now i dont know what we can do, i think we should wait for a response from a maintainer.

**Reply by wu-clan** [reply]
Looking forward to development, but this seems to be a historical legacy, and I also found relevant content in starlette
https://github.com/encode/starlette/pull/2223

**Comment by shaunpatterson** [comment]
Ughhh... bitten by this "bug" today.  This needs to be in the docs.   Why would this be removed?   Quick workaround is to model_dump() then pass the dict into json_encoder
